Dropzone multipart rev2

abannerjee opened the pull request:
Use AWS S3 ManagedUpload in the backend to upload files in multi-part.

Commit:
Add bucketName to DropzoneController.

The bucketName is passed from Bucket to Dropzone and is required
by Dropzone in the paramaters for initiating and completing multipart
uploads.

Commit:
Add function to get the full file path given a folder and file.

This function is added to simplify calculating the 'key' parameter
needed by several aws api calls. The aws 'key' is the full path of
a file on s3.

Commit:
Update the 'updateFormData' function to return a Promise.

Initially, this function was not returning a Promise object and
was throwing an error in the console when running.

This commit updates the function to return a new Promise with
'formData' as the returned variable.

Commit:
Add backend function to initiate multi-part uploading.

A new function is added to AmazonS3Backend to initiate uploading
a part using AWS ManagedUpload. A new ManagedUpload object is
instantiated and attached to the File object which can then be
used to send files to S3.

Commit:
Add backend function to upload file part.

A new function is added to AmazonS3Backend to upload a file
using AWS ManagedUpload. ManagedUpload handles chunking the
file and uploading each part.

This function expects the file and dropzone objects as parameters.
As parts are uploaded, the dropzone progress bar is updated
to reflect upload progress.

The function is named 'uploadPart' to remain universal for
various backends and not just AWS.

Commit:
Override Dropzone uploadFiles method and define accept callback.

The uploadFiles function is overridden to allow uploading files
using the backend 'uploadPart' method instead of using
Dropzone to upload files.

The backend method will allow uploading files in chunks and
will support uploading very large files.

Commit:
Set 'window.onbeforeunload' in the uploadFiles method.

The code to enable prompts when user attempts to navigate away
from the page during upload has been moved out of the Dropzone
'sending' event handler and into the 'uploadFiles' method since
Dropzone is no longer used to send files.

Commit:
Add backend function to cancel uploading files.

A new function is added to AmazonS3Backend to cancel uploading a
file using ManagedUpload. The ManagedUpload abort() function returns
void therefore there is no data to return to the dropzone directive
so the backend function simply returns true.

Commit:
Configure ManagedUpload to compute checksums when uploading parts.

The managed upload object should automatically generate ContentMD5
for each part uploaded.

abannerjee commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Configure Dropzone to use multi-part uploads](https://trello.com/c/iWCrG1wm/3380-configure-dropzone-to-use-multi-part-uploads)

djkiourtsis commented on src/common/AmazonS3Backend.ts:
```diff
@@ -286,19 +286,11 @@ export class AmazonS3Backend implements IBackend {
   public updateFormData(folder: Folder, file: any, formData: any): Promise<any> {
     // append AWS upload key to the form data
     // needed in order to have a valid POST
-    let filePath = file.name;
```
I think it would make sense to change this to use 'getFilePath' in the previous commit.

abannerjee replied:
That would make sense, but also since we are no longer using dropzonejs to submit the form for uploading files, this function never gets called, so I've added a commit to remove 'updateFormData'.

djkiourtsis commented on src/dropzone/DropzoneDirective.ts:
```diff
@@ -47,7 +47,10 @@ export class DropzoneDirective implements ng.IDirective {
         scope.$apply(() => {
           // set form data prior to submitting the dz form
-          scope.$ctrl.backend.updateFormData(scope.$ctrl.folder, file, formData);
+          scope.$ctrl.backend.updateFormData(scope.$ctrl.folder, file, formData)
```
Should we be handling the situation where the promise rejects?  Or is this not needed since we always resolve the promise.

abannerjee replied:
That's right, the 'updateFormData' always resolves and there is never a reject case to handle. Also, same as the previous comment, since we aren't using dropzonejs to upload files, the 'sending' event does not get called. Added a commit removing the 'sending' event handler.

djkiourtsis commented on src/common/AmazonS3Backend.ts:
```diff
@@ -293,4 +293,18 @@ export class AmazonS3Backend implements IBackend {
+  /**
+   * Initiate S3 multi-part upload using ManagedUpload
+   */
+  public initMultipartUpload(params: any): Promise<any> {
```
Using the params dictionary as an argument makes it unclear to me how this works.  Would it make sense to use Bucket, Key, and Body as keyword arguments?  We should also consider passing the folder and file to match the style of updateFormData.

abannerjee replied:
Thanks for pointing this out. AmazonS3Backend implements the IBackend interface, which is supposed to be generic across all backend types. To keep this function generic, it accepts a single params object (since bucket, key, body would be very specific to S3).

I'll update the 'initMultipartUpload' description in AmazonS3Backend to make clear what the expect params object should look like.

djkiourtsis commented on src/common/AmazonS3Backend.ts:
```diff
@@ -293,4 +293,18 @@ export class AmazonS3Backend implements IBackend {
+  public initMultipartUpload(params: any): Promise<any> {
+    return new Promise((resolve: any, reject: any) => {
+      // add ManagedUpload object to the file
+      params.Body.s3upload = new AWS.S3.ManagedUpload({
```
Rather than manually creating a ManagedUpload object, we can use the `this.s3.upload(params, callback)`.  This would also eliminate the need to have a separate function for initializing the multipart upload.

abannerjee replied:
Thanks for pointing this out. I've refactored the code so we do not have two functions for initializing and sending files, but I kept the manual creation of ManagedUpload since we still need to define an event handler on the ManagedUpload object for each file which would listen for 'httpUploadProgress' and update the progress bar in dropzone. These changes are found on the 'rev3' branch. Will be closing this PR.

djkiourtsis commented on src/dropzone/DropzoneDirective.ts:
```diff
@@ -37,9 +37,22 @@ export class DropzoneDirective implements ng.IDirective {
+      accept: acceptCallback
     };
 
+    // dropzone accept
```
Could you elaborate on what this override is doing.

abannerjee replied:
Thanks for pointing this out. This comment has been updated to clearly indicate it is overriding the dropzone accept function to initiate multipart upload on every file upload.

djkiourtsis commented on src/dropzone/DropzoneDirective.ts:
```diff
@@ -92,5 +105,24 @@ export class DropzoneDirective implements ng.IDirective {
       dropzone.on(event, handler);
     });
+
+    Dropzone.prototype.uploadFiles = function(files: any) {
```
Add comments to elaborate on the override.

abannerjee replied:
Thanks for pointing this out, a comment has been added to this override.

djkiourtsis commented on src/dropzone/DropzoneDirective.ts:
```diff
@@ -53,6 +54,17 @@ export class DropzoneDirective implements ng.IDirective {
       });
     }
 
+    // dropzone cancel
```
Add comments to elaborate on the override.

abannerjee replied:
Thanks for pointing this out. This comment has been updated to indicate it is overriding the dropzone cancel function and calling the backend's cancelUpload.

Since we call removeFile after the file successfully uploads (to remove it from the dropzone container), this function always gets called both when the user cancels the file and when it uploads successfully. Therefore, we check to see if the file has actually uploaded fully or not before calling cancelUpload.

Commit:
Remove function updateFormData from backend.

The updateFormData function was previously used to update the
formData by adding the 'key' field needed to make a valid POST to
AWS.

This function is no longer called since file uploads now go through
the backend uploadPart function which uses AWS ManagedUpload to
send the files.

In addition, the Dropzone 'sending' event handler no longer needs
to be defined since Dropzone is not used for sending files, and that
event handler never gets called.

Commit:
Bump s3commander version to 0.8.0.

This version of s3commander supports multi-part uploading to S3.

Commit:
Update README to include ExposeHeader ETag in CORS policy.

In order to use S3 ManagedUpload to upload files to S3, the
S3 bucket needs to include ExposeHeader ETag in the CORS policy.

Commit:
Add additional comments for dropzone overrides and backend functions.

The backend initMultipartUpload function description has been updated
to include the expected params.

The dropzone override comments have been updated to clarify the
new behavior for each function.

abannerjee commented:
Closing this PR in favor of https://github.com/nimbis/s3commander/pull/26
